### PR TITLE
[RISCV] Narrow indices of fixed vector gather/scatter nodes

### PR DESCRIPTION
**Pull request by preames**
Review wise, I'm trying something different.  This PR is what would have been four stacked changes in phabricator.  Having different PRs (and thus duplicated commits) seemed like a major headache, so I'm posting all three changes in one PR.  I will land them individually after review.  Feel free to LGTM individual changes (please be clear on which), and I will rebase as appropriate.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -14225,23 +14259,6 @@ SDValue RISCVTargetLowering::PerformDAGCombine(SDNode *N,
         return DAG.getConstant(-1, DL, VT);
       return DAG.getConstant(0, DL, VT);
     }
-    case Intrinsic::riscv_vloxei:
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:14228]: To draw attention here.  I removed the previous transform over the riscv specific nodes.  I could leave it here if desired, I'm just not sure that's worthwhile.  The result of removing is some test regression in CodeGen/RISCV/rvv/narrow-shift-extend.ll, but I'm not sure if this regression "matters" or if this is synthetic test coverage.  Reverting this piece of the change is simple, I'm happy to go with reviewer preference here.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -11616,15 +11616,44 @@ static SDValue performXORCombine(SDNode *N, SelectionDAG &DAG,
...
 
-  if (N.getOpcode() != ISD::SHL || !N->hasOneUse())
+  if (!N->hasOneUse())
+    return false;
+
+  EVT VT = N.getValueType();
+  SDLoc DL(N);
+
+  // In general, what we're doing here is seeing if we can sink a truncate to
+  // a smaller element type into the expresson tree building our index.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:11632]: expresson -> expression

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -11616,15 +11616,44 @@ static SDValue performXORCombine(SDNode *N, SelectionDAG &DAG,
   return combineSelectAndUseCommutative(N, DAG, /*AllOnes*/ false, Subtarget);
 }
 
-// According to the property that indexed load/store instructions
-// zero-extended their indices, \p narrowIndex tries to narrow the type of index
-// operand if it is matched to pattern (shl (zext x to ty), C) and bits(x) + C <
-// bits(ty).
+/// According to the property that indexed load/store instructions zero-extended
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:11619]: zero-extended -> zero extend

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -11616,15 +11616,44 @@ static SDValue performXORCombine(SDNode *N, SelectionDAG &DAG,
...
+  // a smaller element type into the expresson tree building our index.
+  // TODO: We can generalize this and handle a bunch more cases if useful.
+
+  // Narrow a buildvector to the narrowest element type.  This requires less
+  // work and less register pressure at high LMUL, and creates smaller constants
+  // which may be cheaper to materialize.
+  if (ISD::isBuildVectorOfConstantSDNodes(N.getNode())) {
+    KnownBits Known = DAG.computeKnownBits(N);
+    unsigned ActiveBits = Known.countMaxActiveBits();
+    EVT ResultVT = VT.getVectorElementType();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:11641]: Maybe something like?

`EVT ResultVT = EVT::getIntegerType(Context, ActiveBits).getRoundIntegerType(Context)`?

**Comment by topperc** [comment]
The first 3 patches LGTM

**Comment by preames** [comment]
Since github makes this really hard to follow, let me spell out that the three changes approved by @topperc above were landed as the following:

```
commit 37aa07ad312f58bd2ca59ea712a72544f753f5d3
Author: Philip Reames <preames@rivosinc.com>
Date:   Wed Sep 13 14:56:15 2023 -0700

    [RISCV] Move narrowIndex to be a DAG combine over target

commit 2ff9175af73eb4df7bd2fda35591a95e0c6ba236
Author: Philip Reames <preames@rivosinc.com>
Date:   Thu Sep 14 09:58:52 2023 -0700

    [RISCV] Normalize gather/scatter addressing to UNSIGNED_SCALAR

commit 09a5aac514fa85cb202bfb2faa22ed6b7a29d94a
Author: Philip Reames <preames@rivosinc.com>
Date:   Wed Sep 13 14:39:08 2023 -0700

    [TLI] Add extend as explicit parameter to shouldRemoveExtendFromGSIndex [nfc]

```
I have not yet addressed the changes requested for the final change, and will do so shortly.

**Comment by preames** [comment]
@topperc ping on the last patch

**Review by topperc — APPROVED** [review]
LGTM

**Comment by preames** [comment]
The last change under this PR was landed as e52c55881358e60abbc4811e84ff48eff8f64fbb.
